src/MixedQuoter.sol

```diff
@@ -0,0 +1,311 @@
+                mstore(add(ptr, 0x40), tickAfter)
+                revert(ptr, 0x60)
+            }
+        } else {
+            /// since we don't support exactOutput, revert here
+            revert("Exact output quote not supported");
+        }
+    }
+
+    /// @dev Parses a revert reason that should contain the numeric quote
+    function parseRevertReason(bytes memory reason)
+        private
+        pure
+        returns (uint256 amount, uint160 sqrtPriceX96After, int24 tickAfter)
+    {
+        if (reason.length != 0x60) {
+            if (reason.length < 0x44) revert("Unexpected error");
+            assembly {
+                reason := add(reason, 0x04)
+            }
+            revert(abi.decode(reason, (string)));
+        }
+        return abi.decode(reason, (uint256, uint160, int24));
+    }
+
+    function handleV3Revert(bytes memory reason, IPancakeV3Pool pool, uint256 gasEstimate)
+        private
+        view
+        returns (uint256 amount, uint160 sqrtPriceX96After, uint32 initializedTicksCrossed, uint256)
+    {
+        int24 tickBefore;
+        int24 tickAfter;
+        (, tickBefore,,,,,) = pool.slot0();
+        (amount, sqrtPriceX96After, tickAfter) = parseRevertReason(reason);
+
+        initializedTicksCrossed = pool.countInitializedTicksCrossed(tickBefore, tickAfter);
+
+        return (amount, sqrtPriceX96After, initializedTicksCrossed, gasEstimate);
+    }
+
+    /// @dev Fetch an exactIn quote for a V3 Pool on chain
+    function quoteExactInputSingleV3(QuoteExactInputSingleV3Params memory params)
+        public
+        override
+        returns (uint256 amountOut, uint160 sqrtPriceX96After, uint32 initializedTicksCrossed, uint256 gasEstimate)
+    {
+        bool zeroForOne = params.tokenIn < params.tokenOut;
+        IPancakeV3Pool pool = V3SmartRouterHelper.getPool(factoryV3, params.tokenIn, params.tokenOut, params.fee);
+
+        uint256 gasBefore = gasleft();
+        try pool.swap(
+            address(this), // address(0) might cause issues with some tokens
+            zeroForOne,
+            params.amountIn.toInt256(),
+            params.sqrtPriceLimitX96 == 0
+                ? (zeroForOne ? TickMath.MIN_SQRT_RATIO + 1 : TickMath.MAX_SQRT_RATIO - 1)
+                : params.sqrtPriceLimitX96,
+            abi.encode(params.tokenIn, params.tokenOut, params.fee)
+        ) {} catch (bytes memory reason) {
+            gasEstimate = gasBefore - gasleft();
+            return handleV3Revert(reason, pool, gasEstimate);
+        }
+    }
+
+    /**
+     * V2 *************************************************
+     */
+
+    /// @dev Fetch an exactIn quote for a V2 pair on chain
+    function quoteExactInputSingleV2(QuoteExactInputSingleV2Params memory params)
+        public
+        view
+        override
+        returns (uint256 amountOut)
+    {
+        (uint256 reserveIn, uint256 reserveOut) =
+            V3SmartRouterHelper.getReserves(factoryV2, params.tokenIn, params.tokenOut);
```

Comment:
why is not `V2SmartRouterHelper` 😂

Reply:
got it, can ignore, it's used by both v2 & v3

Reply:
v3 mean copy from v3 repo 😂

Reply:
Same feedback as Burger too, i thought why is v2 related math in v3 😆 

if copied from v3, should we name it the same name eg. `SmartRouterHelper.sol`?

Reply:
actually the SmartRouterHelper is using for v3 SmartRouter , but you guys both think so.
Let me rename this